### PR TITLE
Fix price type

### PR DESCRIPTION
**Pull request by DeliciousHair**
Replacement of structs with public data to ones with private data

Addresses #357 and #365. Although not a very large change, this PR ends up touching rather a lot of code.

* Converted types `Usd`, `Leverage` and `Percent` to something that is appropriate to this application
* Created new types `Price` and `InversePrice` to use for BTC/USD exchange rate with appropriate algebraic ops implemented as well.
* Added new positive tests
* The function `daemon::model::calculate_profit()` has been changed substantially as the updated types make the existing workflow needlessly complex
* Some tests (mostly in `cfd.rs` required updating) in order to make use of the new types.
* Minor edit to `.gitignore` to avoid accidental pushing of DB to repository--should have been it's own item, added here to fix a problem that arose during this work.

NOTE:
* There may be an excess of algebraic ops implemented, some pruning may be appropriate.

### CODE REVIEW

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: Use https://doc.rust-lang.org/std/num/struct.NonZeroU8.html here to avoid this error case.

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: Just a style thing but I don't like `else if` blocks. I would change this to:

```suggestion
        if value.to_f64().unwrap().abs() <= f64::EPSILON {
            return Result::Err(Error::ZeroExchangeRate);
        }
        
        if value.to_f64().unwrap() < 0f64 {
            return Result::Err(Error::NegativeExchangeRate);
        };
```

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: Use https://docs.rs/rust_decimal/1.16.0/rust_decimal/prelude/struct.Decimal.html#method.is_zero instead.

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: Use https://docs.rs/rust_decimal/1.16.0/rust_decimal/prelude/struct.Decimal.html#method.is_sign_negative instead.

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: For `u8`s, I think you should be able to use `From`: https://docs.rs/rust_decimal/1.16.0/rust_decimal/prelude/struct.Decimal.html#impl-From%3Cu8%3E

And that doesn't panic.

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: Formatting should perhaps trim to two decimal places?

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: Do we really want to derive `Serialize` and `Deserialize` here or should we customize the implementation as we did for `Usd`?

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: There is `TryFrom` is you need a fallible conversion trait.

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: These derives will conflict with the manually declared ones!

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: I don't think that is a good practice for the test, better initialize this also with `dec!(123.456)`.

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: There is a constant for 1 BTC.

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs:642]: I would prefer hardcoded expected values over than computed ones.

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: Would be great to use test values that are closer to our real world data.

File: daemon/src/model.rs
Comment (thomaseizinger) [inline daemon/src/model.rs]: This might be more expressive of what is happening here.

```suggestion
    fn usd_for_1_btc_buys_1_btc() {
```

**Review by thomaseizinger — COMMENTED** [review]
Couple of comments to mull over :)

File: daemon/src/model.rs
Comment (DeliciousHair) [reply daemon/src/model.rs]: As an FYI, I actually started with this type but the problem was that except for this error case, all the other features that it offers in terms of our use-case are still experimental and thus in nightly only. So I went back to `u8` for overall convenience.

Can change though, no bid deal either way.

Comment (thomaseizinger) [reply daemon/src/model.rs]: You can use `NonZeroU8` in the fn signature but use a `u8` internally for convenience!

Best of both worlds :)

File: daemon/src/db.rs
Comment (da-kami) [inline daemon/src/db.rs:104]: ❓I'm surprised these type hints are actually needed - can rust not infer the type here based on what is returned?

File: daemon/src/model.rs
Comment (da-kami) [inline daemon/src/model.rs]:
```suggestion
    #[error("Price of zero is not allowed.")]
```

File: daemon/src/model.rs
Comment (da-kami) [inline daemon/src/model.rs]:
```suggestion
    #[error("Negative price is unimplemented.")]
```

File: daemon/src/model.rs
Comment (da-kami) [inline daemon/src/model.rs]: Just to check my understanding of the motivation for `InversePrice`: 
Was `InversePrice` introduced, so we can keep these calculation impls separate from the `Price`? Is there any other motivation for using `InversePrice` over `Price`?

File: daemon/src/model.rs
Comment (da-kami) [inline daemon/src/model.rs]: Please remove this Todo and create an issue for it instead. We should definitely not talk about specific PRs in Todos in the code - once the PR is merged there is no context for this Todo...

File: daemon/src/model.rs
Comment (da-kami) [inline daemon/src/model.rs:654]: 🛠 This test does not test any code at the moment - besides constructing types. If that is intended please rename the test to reflect that.

I think the test should test the actual code where we calculate the liquidation price, not calculate the liquidation price in the test!

File: daemon/src/model/cfd.rs
Comment (da-kami) [inline daemon/src/model/cfd.rs]:
```suggestion
    let is_not_liquidation = closing_price <= long_liquidation_price;
```

File: daemon/src/model/cfd.rs
Comment (da-kami) [inline daemon/src/model/cfd.rs]: 🛠 Hard to understand what `a` and `b` represent - can we somehow come up with better names?

File: daemon/src/model/cfd.rs
Comment (da-kami) [inline daemon/src/model/cfd.rs]: 🛠 Hard to understand what `a` and `b` represent - can we somehow come up with better names?

File: daemon/src/model/cfd.rs
Comment (da-kami) [inline daemon/src/model/cfd.rs]: I could validate that the payout is calculated with the same formula as before.

File: daemon/src/model/cfd.rs
Comment (da-kami) [inline daemon/src/model/cfd.rs]: ❓❗️ Here we are using the long liquidation price (the `test` variable) when defining the bounds for the `Sell` (= `short`) position. Is that correct? 
The liquidation price for the `short` position is substantially higher than the one for `long`. I think the liquidation price calculation has to distinguish between `long` and `short`! Also, the liquidation price for `short` must be guarded by `>= closing_price`.

Or am I misunderstanding something?

Comment (da-kami) [reply daemon/src/model/cfd.rs]: I don't think we should calculate the liquidation price separately in here, but use the liquidation price calculation that already exists (and is hopefully tested for sanity)!

File: daemon/src/model/cfd.rs
Comment (da-kami) [inline daemon/src/model/cfd.rs:1413]: I validated that these values correspond to our GSheet based model. 
For future reference, it would have been great to have a commit, that changes the maintenance margin to `0` prior to removing it - that way you can showcase that the calculation is the same (i.e. tests pass with maintenance margin `0` and without maintenance margin based on the same input and expect).

File: daemon/src/model/cfd.rs
Comment (da-kami) [inline daemon/src/model/cfd.rs]: I think it is sane, but I have a hard time validating that change 😅

**Review by da-kami — CHANGES_REQUESTED** [review]
Great work! This improves understandability a lot!

I have one blocking comment concerning the guard for the `short` `payout` in the `profit` calculation that is based on the `long`'s liquidation price. It might just be a misunderstanding on my side though 😅

Comment on commits: 

Please make sure to clean up the commit history! I think it's good enough to squash the code into one commit. For future work it would be nice to have some clean commits with distinct work, so we can review by commit. 
e.g. for the `.gitignore` file change it would have been great to have a separate commit and a commit message that explains why this change was done in one sentence.

File: daemon/src/model.rs
Comment (DeliciousHair) [reply daemon/src/model.rs]: Nope, none. It's just a safe + convenient way of doing `1/price` is all, since the guardrails are on `Price::new()`

File: daemon/src/model.rs
Comment (DeliciousHair) [reply daemon/src/model.rs]: oops!

File: daemon/src/db.rs
Comment (DeliciousHair) [reply daemon/src/db.rs:104]: Hm, you may be correct... Lemme check that I didn't make extra work here.

EDIT:
Nope, this is "correct" in this context as we are parsing the data to `Decimal` type which is not something we can handle out-of-the-box just yet (see [this issue](https://github.com/launchbadge/sqlx/issues/1489) in particular). This is not to say it cannot be improved, only that this is where we are for the moment.

File: daemon/src/model/cfd.rs
Comment (DeliciousHair) [reply daemon/src/model/cfd.rs]: > Or am I misunderstanding something?

I think you may be; can we please discuss this at some point?

File: daemon/src/model/cfd.rs
Comment (DeliciousHair) [reply daemon/src/model/cfd.rs]: I'm going to say not really, as in this case I'm trying to show what the algebra does, as opposed to how the "concept" works. In this context, using variables `a` and `b` I think make good sense as it removes the context completely, which is the objective in the comment.

File: daemon/src/model/cfd.rs
Comment (DeliciousHair) [reply daemon/src/model/cfd.rs]: Sorry, I'm not sure what you're saying here?

File: daemon/src/model/cfd.rs
Comment (DeliciousHair) [reply daemon/src/model/cfd.rs]: Yeah, I couldn't figure out where the first number was coming from either; suspect there was actually a logic error that was being allowed in via the test, but I really don't know TBH.

File: daemon/src/model/cfd.rs
Comment (DeliciousHair) [reply daemon/src/model/cfd.rs:1413]: Fair enough, but this further ties back to detail that the `maintenance_margin` has now been stripped out completely until it has been properly defined.

File: daemon/src/model/cfd.rs
Comment (DeliciousHair) [reply daemon/src/model/cfd.rs]: > I don't think we should calculate the liquidation price separately in here, but use the liquidation price calculation that already exists (and is hopefully tested for sanity)!

The existing function is:
```rust
fn calculate_liquidation_price(leverage: Leverage, price: Price) -> Price {
    price * leverage / (leverage + 1)
}
```
which I wanted to remove TBH as having such trivial calculations relegated out to distinct functions makes the code harder to read IMHO as I the maintainer have to navigate around to discover that a function that does something akin to `a + b` has been used at some location rather than just writing `a + b`.

If the math was more complicated, sure! But in this case, I really think it is just making things less clear.

Happy enough to go either way, but will make consistent.

Also, re: testing please note from `model.rs` that:
```rust
#[test]
fn test_liquidation() {
    let usd = Usd::new(dec!(61234.5678));
    let leverage = Leverage::new(5).unwrap();
    let price = Price::new(dec!(61234.5678)).unwrap();
    let expected_buyin = Amount::from_str_in("0.2", Denomination::Bitcoin).unwrap();

    let liquidation_price = price * leverage / (leverage + 1);
    let inv_price = InversePrice::new(price);
    let inv_liquidation_price = InversePrice::new(liquidation_price);

    let long_buyin = usd / (price * leverage);
    let long_payout =
        (usd / leverage) * ((leverage + 1) * inv_price - leverage * inv_liquidation_price);

    assert_eq!(long_buyin, expected_buyin);
    assert_eq!(long_payout, Amount::ZERO);
}
```

File: .gitignore
Comment (klochowicz) [inline .gitignore]:
```suggestion
```

Comment (klochowicz) [reply .gitignore]: nit: I believe you don't need this one, if we're ignoring all the sqlite databases in the repo with your new change

File: daemon/src/model.rs
Comment (klochowicz) [inline daemon/src/model.rs]: could we please add a check that `value.0` is not `ZERO`? (I know that currently price can't be created with zero, but it could be a footgun if we relax that requirement or perform calculations resulting in zero.

File: daemon/src/model.rs
Comment (klochowicz) [inline daemon/src/model.rs]: Perhaps we could create a ticket instead of TODO? (I think we agreed on that during last retro)

File: daemon/src/model/cfd.rs
Comment (klochowicz) [inline daemon/src/model/cfd.rs]: Please remove the empty lines here as it interferes with `rustfmt` (@thomaseizinger always points it out in my reviews 😄 )

(here and below)

File: daemon/src/model/cfd.rs
Comment (klochowicz) [inline daemon/src/model/cfd.rs:161]: what happened to `maintenance_margin_rate`? It seems odd that it was removed as part of this PR. If you'd like to remove it, I'd prefer to open up a separate PR so we can focus and discuss it, instead of blocking the refactor

**Review by klochowicz — COMMENTED** [review]
it's getting close, I guess the commits should be squashed with some nice description before it can get merged.

File: daemon/src/model.rs
Comment (DeliciousHair) [reply daemon/src/model.rs:654]: Sorry, but I disagree. The types are in fact being tested in the sense that we are creating some values and doing some algebra. The fact that the algebra is testing that the liquidation price yields a payout of zero is merely a convenient test value.

And again, computing the liquidation price is trivial--the function in another module is the exact same one-line operation.

As for renaming the test, alright.

File: daemon/src/model/cfd.rs
Comment (da-kami) [reply daemon/src/model/cfd.rs]: I am always comparing what we have in code with the model that we have in Google sheets. That way I can somehow verify that we have the expected outcome (besides the tests). I just added those comments so that other reviewers might see that it was verified against the GSheet model. that what @bonomat and me did with previous PRs that changed calculations.

File: daemon/src/model.rs
Comment (DeliciousHair) [reply daemon/src/model.rs]: Since `Price.0 = ZERO` is blocked, I don't  see what you're getting at here as the check would serve no purpose.

Comment (klochowicz) [reply daemon/src/model.rs]: I mentioned above -e - if anyone ever decides to support price being `0` or be able to calculate it (as a result of some computation, not via `new()`), then this will be an error waiting to happen.
It's much easier to be explicit here, instead of forcing a reader to put trust that these types will be set in lockstep. 

Is there any harm in being a bit more explicit here stating the contract?

File: daemon/src/model/cfd.rs
Comment (DeliciousHair) [reply daemon/src/model/cfd.rs]: Given that this is a comment only and is written up as formulae, I think the use of the single-letter values is appropriate within the context.

File: daemon/src/model/cfd.rs
Comment (DeliciousHair) [reply daemon/src/model/cfd.rs]: Ah, gotcha!

File: daemon/src/model/cfd.rs
Comment (DeliciousHair) [reply daemon/src/model/cfd.rs]: Fixed with current update, should all be well now!

File: daemon/src/model/cfd.rs
Comment (DeliciousHair) [reply daemon/src/model/cfd.rs:161]: It was simply dropped as it was ad-hoc'd into things as the pedagogy of said item relied on some cumbersome algebra. With the current structure it is easy to compute on the fly in both the payout curve and the profit calculation(s), or can be stuck onto, say, the `Cfd` struct as a static value--doesn't really matter TBH; note that the resulting functions in `cfd.rs` are simply:

```rust
pub fn calculate_buy_margin(price: Price, quantity: Usd, leverage: Leverage) -> Amount {
    quantity / (price * leverage)
}

fn calculate_sell_margin(price: Price, quantity: Usd) -> Amount {
    quantity / price
}

fn calculate_liquidation_price(leverage: Leverage, price: Price) -> Price {
    price * leverage / (leverage + 1)
}
```

File: daemon/src/model.rs
Comment (DeliciousHair) [reply daemon/src/model.rs]: I suppose not, so added it is!

File: daemon/src/model.rs
Comment (da-kami) [reply daemon/src/model.rs]: It's still a Todo, can we move to a ticket?

File: daemon/src/model.rs
Comment (da-kami) [inline daemon/src/model.rs]:
```suggestion
    fn test_algebra() {
```

Reasoning: This test is not about testing the liquidation price but just algebra and constructing types. The liquidation price is just used as an example, but no code calculating the liquidation price is tested.

**Review by da-kami — APPROVED** [review]
nice, thanks!

File: daemon/src/model.rs
Comment (DeliciousHair) [reply daemon/src/model.rs]: That is actually (more or less) what I had and was told to change it to (more or less) the name that is there.  I'll change it again as I prefer your suggestion, but perhaps we should create a style-guide for this sort of thing as there is clearly some inconsistency happening?

File: daemon/src/model/cfd.rs
Comment (klochowicz) [reply daemon/src/model/cfd.rs:161]: This explanation makes sense, thank you! Would you mind mentioning it also in the commit message? otherwise people might be wondering why it went away during a refactor.

**Review by klochowicz — APPROVED** [review]

**Pull request merged**
